---
title: "DESeq Analysis on Teee RNA-Seq"
output:
  pdf_document: 
    toc: true
  html_document: 
    toc: true
date: "2022-08-17"
---

```{r library_loading, message=FALSE}
library(Rsubread)
library(DESeq2)
library(AnnotationDbi)
library(org.Hs.eg.db)
library(EnhancedVolcano)
library(clusterProfiler)
library(tidyverse)
library(pheatmap)
library(VennDiagram)
library(ggfortify)
library(WGCNA)
```

# Load and clean counts data

```{r count_data_loading, message=FALSE}
counts_df = readr::read_table("RNA_Seq_counts.out", comment = "#", skip_empty_rows = TRUE)
counts_df = data.frame(counts_df)

names(counts_df) = c("Gene_ID", 
                    "Chr", 
                    "Start", 
                    "End", 
                    "Strand", 
                    "Length", 
                    "Sample_1_CD38_neg_CD39_neg",
                    "Sample_1_CD38_neg_CD39_pos",
                    "Sample_1_CD38_pos_CD39_neg",
                    "Sample_1_CD38_pos_CD39_pos",
                    "Sample_2_CD38_neg_CD39_neg",
                    "Sample_2_CD38_neg_CD39_pos",
                    "Sample_2_CD38_pos_CD39_neg",
                    "Sample_2_CD38_pos_CD39_pos",
                    "Sample_3_CD38_neg_CD39_neg",
                    "Sample_3_CD38_neg_CD39_pos",
                    "Sample_3_CD38_pos_CD39_neg",
                    "Sample_3_CD38_pos_CD39_pos")

#remove extra columns
rownames(counts_df) = counts_df$Gene_ID
counts_df = counts_df[,c(7:18)]

#remove NA rows
counts_df = tidyr::drop_na(counts_df)

#remove 0 count rows
counts_df = counts_df[rowSums(counts_df) > 50, ]

#need to write a function to filter out rows with less than a certain number of cells in a certain proportion of samples.##

```

# Create a sample info dataframe

```{r sample_info}

population = factor(c("neg_neg", "neg_pos", "pos_neg", "pos_pos", "neg_neg", "neg_pos", "pos_neg", "pos_pos", "neg_neg", "neg_pos", "pos_neg", "pos_pos"))

patient = factor(c("Patient_01", "Patient_01", "Patient_01", "Patient_01", "Patient_02", "Patient_02", "Patient_02", "Patient_02", "Patient_03", "Patient_03", "Patient_03", "Patient_03"))

#batch = factor(c("Batch_01", "Batch_01", "Batch_01", "Batch_01", "Batch_02", "Batch_02", "Batch_02", "Batch_02", "Batch_02", "Batch_02", "Batch_02", "Batch_02"))

condition_df = data.frame(row.names = colnames(counts_df), population, patient)

rm(list = c("patient", "population", "batch"))

```

# Create a DESeq object
```{r DESeq, message=FALSE}

dds = DESeq2::DESeqDataSetFromMatrix(countData = counts_df, 
                             colData = condition_df, 
                             design = ~ patient + population)
dds = DESeq2::DESeq(dds)

plotDispEsts(dds)

```

# Create normalized expression data frames
```{r Normalized data frame, message=FALSE}

#make a normalized gene expression dataframe
normalized_expression_df = as.data.frame(counts(dds, normalized = TRUE))

#write to csv
write.csv(normalized_expression_df, file = "RNA_Seq_normalized_counts.csv")

#Add gene names
normalized_expression_df$gene = mapIds(org.Hs.eg.db, 
                                       keys=rownames(normalized_expression_df), 
                                       keytype = "ENSEMBL",
                                       column = "SYMBOL")

#Make a transposed data frame
df_transposed = as.data.frame(t(normalized_expression_df))
df_transposed = rownames_to_column(df_transposed, "Sample_ID")
df_transposed = df_transposed[1:12,]
df_transposed$population = condition_df$population
df_transposed$patient = condition_df$patient
df_transposed = select(df_transposed, Sample_ID, patient, population, everything())
df_transposed[,4:ncol(df_transposed)] = sapply(df_transposed[,4:ncol(df_transposed)], as.numeric)
```


# DESeq contrasts 
```{r DESeq contrasts}

#contrasts
results_pospos_vs_negneg = results(dds, contrast = c("population", "pos_pos", "neg_neg"))
results_pospos_vs_posneg = results(dds, contrast = c("population", "pos_pos", "pos_neg"))
results_pospos_vs_negpos = results(dds, contrast = c("population", "pos_pos", "neg_pos"))

#the results function can do all contrasts together

#Add gene names
results_pospos_vs_negneg$gene = mapIds(org.Hs.eg.db, 
                                       keys=rownames(results_pospos_vs_negneg), 
                                       keytype = "ENSEMBL",
                                       column = "SYMBOL")
results_pospos_vs_posneg$gene = mapIds(org.Hs.eg.db, 
                                       keys=rownames(results_pospos_vs_posneg), 
                                       keytype = "ENSEMBL",
                                       column = "SYMBOL")
results_pospos_vs_negpos$gene = mapIds(org.Hs.eg.db, 
                                       keys=rownames(results_pospos_vs_negpos), 
                                       keytype = "ENSEMBL",
                                       column = "SYMBOL")

#summary
summary(results_pospos_vs_negneg)
summary(results_pospos_vs_posneg)
summary(results_pospos_vs_negpos)

#remove NAs, filter for adjusted p-value of <0.01, and Log2FC of > 1
sig_pospos_vs_negneg = na.omit(results_pospos_vs_negneg)
sig_pospos_vs_negneg = sig_pospos_vs_negneg[sig_pospos_vs_negneg$padj < 0.01, ]
sig_pospos_vs_negneg = sig_pospos_vs_negneg[abs(sig_pospos_vs_negneg$log2FoldChange) > 1 , ]

sig_pospos_vs_posneg = na.omit(results_pospos_vs_posneg)
sig_pospos_vs_posneg = sig_pospos_vs_posneg[sig_pospos_vs_posneg$padj < 0.01, ]
sig_pospos_vs_posneg = sig_pospos_vs_posneg[abs(sig_pospos_vs_posneg$log2FoldChange) > 1 , ]

sig_pospos_vs_negpos = na.omit(results_pospos_vs_negpos)
sig_pospos_vs_negpos = sig_pospos_vs_negpos[sig_pospos_vs_negpos$padj < 0.01, ]
sig_pospos_vs_negpos = sig_pospos_vs_negpos[abs(sig_pospos_vs_negpos$log2FoldChange) > 1 , ]

#view significant results
sig_pospos_vs_negneg
sig_pospos_vs_posneg
sig_pospos_vs_negpos

#overlapping DEGs
negneg_negpos_overlap = merge(as.data.frame(sig_pospos_vs_negneg), as.data.frame(sig_pospos_vs_negpos), by="gene")
negneg_posneg_overlap = merge(as.data.frame(sig_pospos_vs_negneg), as.data.frame(sig_pospos_vs_posneg), by="gene")
posneg_negpos_overlap = merge(as.data.frame(sig_pospos_vs_posneg), as.data.frame(sig_pospos_vs_negpos), by="gene")
overlap_list = list(negneg_negpos_overlap, negneg_posneg_overlap, posneg_negpos_overlap)
all_overlap = overlap_list %>% 
  reduce(merge, by = "gene") 
matrix_overlapping_gene = merge(normalized_expression_df, all_overlap, by = "gene")
matrix_overlapping_gene = matrix_overlapping_gene[,1:13]
rm(overlap_list)
rm(all_overlap)

```

# Visualize data

## Histograms of p-value distributions
```{r p-value histograms, fig.width=4, fig.height=4, dpi=300}
#histograms of unadjusted p-values
hist(results_pospos_vs_negneg$pvalue)
hist(results_pospos_vs_posneg$pvalue)
hist(results_pospos_vs_negpos$pvalue)
```

## MA plots
```{r MA plots, fig.width=5, fig.height=5, dpi=200}
#MA plots
plotMA(results_pospos_vs_negneg, alpha = 0.01, main = "CD38+CD39+ vs. CD38-CD39-", colSig = "red")
plotMA(results_pospos_vs_posneg, alpha = 0.01, main = "CD38+CD39+ vs. CD38+CD39-", colSig = "red")
plotMA(results_pospos_vs_negpos, alpha = 0.01, main = "CD38+CD39+ vs. CD38-CD39+", colSig = "red")
```
## PCA
```{r PCA, fig.width=5, fig.height=4, dpi=200}
#PCA
pca_results = prcomp(df_transposed[,4:ncol(df_transposed)], center = TRUE, scale. = TRUE)

#pdf("PCA.pdf", width = 5.5, height = 4)
autoplot(pca_results, data = df_transposed, colour = "population", shape = "patient", size = 6) +
  scale_color_manual(values = c("black", "blue", "orange", "red")) +
  #geom_rug(aes(colour = population), size = 2, outside = TRUE, sides = "b", length = unit(0.01, "npc")) +
  coord_cartesian(clip = "off") +
  theme_bw()
#dev.off()

#Loadings plot
pc1 = as.data.frame(pca_results$rotation[,1])
pc1$absolute = abs(pc1$`pca_results$rotation[, 1]`)
pc1 = dplyr::arrange(pc1, by = absolute)

barplot(pc1[1:20,1], main="PC 1 Loadings Plot")

#DESeq2 PCA
vsd = vst(dds, blind=FALSE)
plotPCA(vsd, intgroup="population")


```

## Volcano plots

```{r Volcano plots, fig.width=6, fig.height=6, dpi=200}

#Volcano plots

#pdf("Volcano_plots.pdf", width = 10, height = 10)
EnhancedVolcano(results_pospos_vs_negneg, 
                x="log2FoldChange", 
                y="padj", 
                title="CD38+CD39+ vs. CD38-CD39-",
                lab=results_pospos_vs_negneg$gene,
                pCutoff = 0.01,
                FCcutoff = 1.5,
                col = c("grey", "grey", "grey", "red"),
                colAlpha = 1)

EnhancedVolcano(results_pospos_vs_posneg, 
                x="log2FoldChange", 
                y="padj", 
                title="CD38+CD39+ vs. CD38+CD39-",
                lab=results_pospos_vs_posneg$gene,
                pCutoff = 0.01,
                FCcutoff = 1.5,
                col = c("grey", "grey", "grey", "red"),
                colAlpha = 1,
                xlim = c(-45,45)
                )

EnhancedVolcano(results_pospos_vs_negpos, 
                x="log2FoldChange", 
                y="padj", 
                title="CD38+CD39+ vs. CD38-CD39+",
                lab=results_pospos_vs_negpos$gene,
                pCutoff = 0.01,
                FCcutoff = 1.5,
                col = c("grey", "grey", "grey", "red"),
                colAlpha = 1,
                xlim = c(-25,25)
                )
#dev.off()

```
## Heatmap

```{r Heatmap, fig.width=7, fig.height=10, dpi=200}

#heatmap of overlapping genes

#pdf("Overlap_heatmap.pdf", width = 5, height = 10)
pheatmap(matrix_overlapping_gene[,2:13], scale = "row", fontsize_col = 8, fontsize_row = 7, labels_row = matrix_overlapping_gene$gene)
#dev.off()

```

## Venn Diagram

```{r Venn_Diagram, dpi=200, fig.height=5, fig.width=5}

#vs 38-39- : 809 significant genes 809
#vs 38+39- : 604 significant genes 604
#vs 38-39+ : 168 significant genes 168

#overlapping in both 38-39- and 38+39-: 397
#overlapping in both 38-39- and 38-39+: 131
#overlapping in both 38-39+ and 38+39-: 100

#overlapping in all: 90

#pdf("Venn_Diagram.pdf", width = 4, height = 4)
grid.newpage()
draw.triple.venn(area1=809, area2=604, area3=168, n12=397, n23=100, n13=131,n123=90, 
                 category=c('vs. CD38-CD39-', 'vs. CD38+CD39-', 'vs. CD38-CD39+'),  # label for each circle
                 fill=c('black', 'orange', 'blue'), # color for each circle
                 lwd=c(0,0,0)  # line width for each circle 
                  )
#dev.off()

```

## Plotting individual genes

```{r Gene_Plots, fig.height=3, fig.width=5, dpi=200}

#CXCR6
ggplot(df_transposed, aes(x=population, y=as.numeric(ENSG00000172215), shape=patient, color=patient)) + 
  geom_jitter(size = 10, width=0.2) +
  ylab(label= "Normalized Counts") +
  ggtitle(label = "CXCR6") +
  scale_color_manual(values = c("red", "blue", "green")) +
  theme_classic() +
  theme(text=element_text(size=24))

```

# Plotting function
```{r}

##make a plotting function and feed a list of genes##
ENSG_plot = function(ENSG) {
  group <- "patient"
  pop <- "population"
  plt <- ggplot(df_transposed, aes_(x=as.name(pop), y=as.name(ENSG), shape=as.name(group), color=as.name(group))) +
    geom_jitter(size = 6, width=0.2) +
    ylab(label= "Normalized Counts") +
    ggtitle(label = {{ENSG}}) +
    scale_color_manual(values = c("red", "blue", "green")) +
    theme_classic() +
    theme(text=element_text(size=18))
  return(plt)
}
```

# Plots sets of genes
```{r}
#Proliferation genes

CDK_genes = c("ENSG00000148773", "ENSG00000170312", "ENSG00000123374",  "ENSG00000135446", "ENSG00000105810", "ENSG00000110092", "ENSG00000118971", "ENSG00000112576", "ENSG00000105173", "ENSG00000175305", "ENSG00000133101", "ENSG00000145386")

#ENSG00000148773 = Ki67
#ENSG00000170312 = CDK1s
#ENSG00000123374 = CDK2
#ENSG00000250506 = CDK3 (NOT IN DATA)
#ENSG00000135446 = CDK4
#ENSG00000105810 = CDK6
#ENSG00000110092 = CCND1
#ENSG00000118971 = CCND2
#ENSG00000112576 = CCND3
#ENSG00000105173 = CCNE1
#ENSG00000175305 = CCNE2
#ENSG00000133101 = CCNA1
#ENSG00000145386 = CCNA2

for(gene in CDK_genes) {
  print(ENSG_plot(gene))
}

#Polarization genes

polarization_genes = c("ENSG00000073861", "ENSG00000163508", "ENSG00000107485", "ENSG00000049768", "ENSG00000143365", "ENSG00000111537", "ENSG00000232810", "ENSG00000113520", "ENSG00000105329", "ENSG00000127743", "ENSG00000124391", "ENSG00000138684", "ENSG00000110944", "ENSG00000109471", "ENSG00000136634", "ENSG00000100453", "ENSG00000113088", "ENSG00000145649", "ENSG00000197540")

#IL-17A not in data
#ENSG00000073861 = TBX21
#ENSG00000163508 = EOMES
#ENSG00000107485 = GATA3
#ENSG00000049768 = FOXP3
#ENSG00000143365 = RORC
#ENSG00000111537 = IFNG
#ENSG00000232810 = TNF
#ENSG00000113520 = IL4
#ENSG00000105329 = TGFB1
#ENSG00000092969 = TGFB2 (NOT IN DATA)
#ENSG00000136244 = IL6 (NOT IN DATA)
#ENSG00000145839 = IL9 (NOT IN DATA)
#ENSG00000127743 = IL17B
#ENSG00000124391 = IL17C
#ENSG00000112116 = IL17F (NOT IN DATA)
#ENSG00000138684 = IL21
#ENSG00000110944 = IL23A
#ENSG00000109471 = IL2
#ENSG00000113525 = IL5 (NOT IN DATA)
#ENSG00000169194 = IL13 (NOT IN DATA)
#ENSG00000136634 = IL10
#ENSG00000100453 = GZMB
#ENSG00000113088 = GZMK
#ENSG00000145649 = GZMA
#ENSG00000197540 = GZMM


for(gene in polarization_genes) {
  print(ENSG_plot(gene))
}

#Checkpoint genes

checkpoints_to_plot = c("ENSG00000120217", "ENSG00000197646", "ENSG00000188389", "ENSG00000163599", "ENSG00000135077", "ENSG00000089692", "ENSG00000178562", "ENSG00000186265", "ENSG00000157873", "ENSG00000117281", "ENSG00000168961", "ENSG00000181847", "ENSG00000186891", "ENSG00000163600", "ENSG00000139193")

#B7-H4 isn't present in data
#ENSG00000120217 = CD274
#ENSG00000197646 = PDCD1LG2
#ENSG00000188389 = PDCD1
#ENSG00000163599 = CTLA4
#ENSG00000135077 = HAVCR2
#ENSG00000089692 = LAG3
#ENSG00000178562 = CD28
#ENSG00000186265 = BTLA
#ENSG00000157873 = TNFRSF14 (HVEM)
#ENSG00000117281 = CD160
#ENSG00000168961 = GAL9
#ENSG00000181847 = TIGIT
#ENSG00000186891 = GITR (TNFRSF18)
#ENSG00000163600 = ICOS
#ENST00000407780 = ICOSL (NOT IN DATA)
#ENST00000281834 = OX40 (NOT IN DATA)
#ENST00000370629 = CD40L (NOT IN DATA)
#ENSG00000139193 = CD27

for(gene in checkpoints_to_plot) {
  print(ENSG_plot(gene))
}


```
#Function for normalizing of paired data
```{r}
normalize_by_group_sum <- function(df, group_by){
  # Arguments: df: dataframe with numeric columns to normalize 
  #    by group sum
  # group_by: str column name to group by 
  # returns dataframe with all numeric cols divided by group sum (float type)
  
  numeric_cols <- colnames(df)[unlist(lapply(df, is.numeric))]
  
  df_grp <- df %>%
    group_by(patient) %>%
    summarize(across(where(is.numeric), sum, na.rm=TRUE))
  
  # merge
  df_comb <- inner_join(df, df_grp, by="patient")
  
  # for each numeric .x column, divide e.g. CASP3.x by CASP3.y
  numeric_cols.x <- paste0(numeric_cols, ".x")
  numeric_cols.y <- paste0(numeric_cols, ".y")
  
  # for loop
  for (col.x in numeric_cols.x) {
    newcol <- gsub('.x','', col.x)
    col.y <- gsub('.x', '.y', col.x)
    df_comb[[newcol]] <- df_comb[[col.x]] / df_comb[[col.y]]
  }
  
  # now keep only the non-x/y columns which were used for interim calcs
  df_comb <- df_comb %>% 
    select (-c(numeric_cols.x, numeric_cols.y))
  
  return (df_comb)
}
```

# Heatmaps of selected gene sets
```{r}
#Apoptosis

#pdf("selected_heatmaps.pdf", width = 6, height = 4)

apoptosis_genes = c("ENSG00000164305", "ENSG00000165806", "ENSG00000064012", "ENSG00000132906", "ENSG00000171791", "ENSG00000087088", "ENSG00000171552", "ENSG00000141510", "ENSG00000143384")

apoptosis_df = df_transposed %>% 
  #only select columns for genes of interest
  select(patient, population, all_of(apoptosis_genes)) %>% 
  #rename the ensembl ids to gene names
  data.table::setnames(old = apoptosis_genes, new = c("CASP3", "CASP7", "CASP8", "CASP9", "BCL2", "BAX", "BCL-XL", "TP53", "MCL1" )) %>% 
  #merge population name and patient ID
  unite(col = "ID", patient, population, remove = FALSE) %>% 
  #sort based on population
  arrange(population) %>% 
  #normalizing
  normalize_by_group_sum("patient")

pheatmap(apoptosis_df[,4:ncol(apoptosis_df)], scale = "column", labels_row = apoptosis_df$ID, main = "Apoptosis Genes")


#Proliferation

CDK_genes = c("ENSG00000148773", "ENSG00000170312", "ENSG00000123374",  "ENSG00000135446", "ENSG00000105810", "ENSG00000110092", "ENSG00000118971", "ENSG00000112576", "ENSG00000105173", "ENSG00000175305", "ENSG00000133101", "ENSG00000145386")

CDK_df = df_transposed %>% 
  #only select columns for genes of interest
  select(patient, population, all_of(CDK_genes)) %>% 
  #rename the ensembl ids to gene names
  data.table::setnames(old = CDK_genes, new = c("MKI67", "CDK1", "CDK2", "CDK4", "CDK6", "CCND1", "CCND2", "CCND3", "CCNE1", "CCNE2", "CCNA1", "CCNA2" )) %>% 
  #merge population name and patient ID
  unite(col = "ID", patient, population, remove = FALSE) %>% 
  #sort based on population
  arrange(population) %>% 
  #normalize
  normalize_by_group_sum("patient")

pheatmap(CDK_df[,4:ncol(CDK_df)], scale = "column", labels_row = CDK_df$ID, main = "Cell Cycle Genes")

#Polarization

polarization_genes = c("ENSG00000073861", "ENSG00000163508", "ENSG00000107485", "ENSG00000049768", "ENSG00000143365", "ENSG00000111537", "ENSG00000232810", "ENSG00000113520", "ENSG00000105329", "ENSG00000127743", "ENSG00000124391", "ENSG00000138684", "ENSG00000110944", "ENSG00000109471", "ENSG00000136634", "ENSG00000100453", "ENSG00000113088", "ENSG00000145649", "ENSG00000197540")

polarization_df = df_transposed %>% 
  #only select columns for genes of interest
  select(patient, population, all_of(polarization_genes)) %>% 
  #rename the ensembl ids to gene names
  data.table::setnames(old = polarization_genes, new = c("TBX21", "EOMES", "GATA3", "FOXP3", "RORC", "IFNG", "TNF", "IL4", "TGFB1", "IL17B", "IL17C", "IL21", "IL23A", "IL2", "IL10", "GZMB", "GZMK", "GZMA", "GZMM")) %>% 
  #merge population name and patient ID
  unite(col = "ID", patient, population, remove = FALSE) %>% 
  #sort based on population
  arrange(population) %>% 
  #normalize
  normalize_by_group_sum("patient")

pheatmap(polarization_df[,4:ncol(polarization_df)], scale = "column", labels_row = polarization_df$ID,  main = "Polarization Genes")

#Checkpoints
checkpoint_genes = c("ENSG00000120217", "ENSG00000197646", "ENSG00000188389", "ENSG00000163599", "ENSG00000135077", "ENSG00000089692", "ENSG00000178562", "ENSG00000186265", "ENSG00000157873", "ENSG00000117281", "ENSG00000168961", "ENSG00000181847", "ENSG00000186891", "ENSG00000163600", "ENSG00000139193")

checkpoint_df = df_transposed %>% 
  #only select columns for genes of interest
  select(patient, population, all_of(checkpoint_genes)) %>% 
  #rename the ensembl ids to gene names
  data.table::setnames(old = checkpoint_genes, new = c("CD274", "PDCD1LG2", "PDCD1", "CTLA4", "HAVCR2", "LAG3", "CD28", "BTLA", "TNFRSF14", "CD160", "GAL9", "TIGIT", "TNFRSF18", "ICOS", "CD27")) %>% 
  #merge population name and patient ID
  unite(col = "ID", patient, population, remove = FALSE) %>% 
  #sort based on population
  arrange(population) %>% 
  #normalize
  normalize_by_group_sum("patient")

pheatmap(checkpoint_df[,4:ncol(checkpoint_df)], scale = "column", labels_row = checkpoint_df$ID, main = "Checkpoint Genes")

#T-cell migration
migration_genes = c("ENSG00000002586", "ENSG00000124343", "ENSG00000112486", "ENSG00000161921", "ENSG00000169508",  "ENSG00000115009", "ENSG00000271503", "ENSG00000136286", "ENSG00000090339", "ENSG00000005844", "ENSG00000115232", "ENSG00000099942", "ENSG00000113811", "ENSG00000064651", "ENSG00000157873", "ENSG00000180549", "ENSG00000158769", "ENSG00000154165", "ENSG00000147065", "ENSG00000139626", "ENSG00000102879", "ENSG00000186810")

migration_df = df_transposed %>% 
  #only select columns for genes of interest
  select(patient, population, all_of(migration_genes)) %>% 
  #rename the ensembl ids to gene names
  data.table::setnames(old = migration_genes, new = c("CD99", "XG", "CCR6", "CXCL16", "gpr183",  "CCL20", "CCL5", "Myo1g", "ICAM1", "Itgal", "Itga4", "Crkl", "Selenok", "Slc12a2", "Tnfrsf14", "Fut7", "F11r", "Gpr15", "Msn", "Itgb7", "Coro1a", "CXCR3")) %>% 
  #merge population name and patient ID
  unite(col = "ID", patient, population, remove = FALSE) %>% 
  #sort based on population
  arrange(population) %>% 
  #normalize
  normalize_by_group_sum("patient")

pheatmap(migration_df[,4:ncol(migration_df)], scale = "column", labels_row = migration_df$ID, main = "Checkpoint Genes")

#dev.off()

```

# GSEA

```{r GSEA}

#CD38+CD39+ vs CD38-CD39-
pospos_vs_negneg_for_gsea = results_pospos_vs_negneg$log2FoldChange
names(pospos_vs_negneg_for_gsea) = rownames(results_pospos_vs_negneg)
pospos_vs_negneg_for_gsea = sort(pospos_vs_negneg_for_gsea, decreasing = TRUE)

gse_pospos_vs_negneg = gseGO(geneList = pospos_vs_negneg_for_gsea,
            ont = "BP",
            keyType = "ENSEMBL",
            OrgDb = "org.Hs.eg.db")

head(gse_pospos_vs_negneg, n=25)

#CD38+CD39+ vs CD38+CD39-
pospos_vs_posneg_for_gsea = results_pospos_vs_posneg$log2FoldChange
names(pospos_vs_posneg_for_gsea) = rownames(results_pospos_vs_posneg)
pospos_vs_posneg_for_gsea = sort(pospos_vs_posneg_for_gsea, decreasing = TRUE)

gse_pospos_vs_posneg = gseGO(geneList = pospos_vs_posneg_for_gsea,
            ont = "BP",
            keyType = "ENSEMBL",
            OrgDb = "org.Hs.eg.db")

head(gse_pospos_vs_posneg, n=25)

#CD38+CD39+ vs CD38-CD39+
pospos_vs_negpos_for_gsea = results_pospos_vs_negpos$log2FoldChange
names(pospos_vs_negpos_for_gsea) = rownames(results_pospos_vs_negpos)
pospos_vs_negpos_for_gsea = sort(pospos_vs_negpos_for_gsea, decreasing = TRUE)

gse_pospos_vs_negpos = gseGO(geneList = pospos_vs_negpos_for_gsea,
            ont = "BP",
            keyType = "ENSEMBL",
            OrgDb = "org.Hs.eg.db")

head(gse_pospos_vs_negpos, n=25)

#plotting GSEA results
#gseaplot(gse_pospos_vs_negneg, geneSetID = 1)

```

# GO Analysis
```{r GO Analysis, fig.height=4, fig.width=7, dpi=200}

#pospos_vs_negneg increased genes
GO_up_pospos_vs_negneg_genes = rownames(sig_pospos_vs_negneg[sig_pospos_vs_negneg$log2FoldChange > 1,])
GO_up_pospos_vs_negneg_results = enrichGO(gene = GO_up_pospos_vs_negneg_genes, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
summary(GO_up_pospos_vs_negneg_results)
plot(barplot(GO_up_pospos_vs_negneg_results, showCategory = 10))

#pospos_vs_negneg decreased genes
GO_down_pospos_vs_negneg_genes = rownames(sig_pospos_vs_negneg[sig_pospos_vs_negneg$log2FoldChange < 1,])
GO_down_pospos_vs_negneg_results = enrichGO(gene = GO_down_pospos_vs_negneg_genes, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
summary(GO_down_pospos_vs_negneg_results)
#plot(barplot(GO_down_pospos_vs_negneg_results, showCategory = 10))
##NO SIG PATHWAYS

#pospos_vs_posneg increased genes
GO_up_pospos_vs_posneg_genes = rownames(sig_pospos_vs_posneg[sig_pospos_vs_posneg$log2FoldChange > 1,])
GO_up_pospos_vs_posneg_results = enrichGO(gene = GO_up_pospos_vs_posneg_genes, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
summary(GO_up_pospos_vs_posneg_results)
plot(barplot(GO_up_pospos_vs_posneg_results, showCategory = 10))

#pospos_vs_posneg decreased genes
GO_down_pospos_vs_posneg_genes = rownames(sig_pospos_vs_posneg[sig_pospos_vs_posneg$log2FoldChange < 1,])
GO_down_pospos_vs_posneg_results = enrichGO(gene = GO_down_pospos_vs_posneg_genes, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
summary(GO_down_pospos_vs_posneg_results)
#plot(barplot(GO_down_pospos_vs_posneg_results, showCategory = 10))
##NO SIG PATHWAYS

#pospos_vs_negpos increased genes
GO_up_pospos_vs_negpos_genes = rownames(sig_pospos_vs_negpos[sig_pospos_vs_negpos$log2FoldChange > 1,])
GO_up_pospos_vs_negpos_results = enrichGO(gene = GO_up_pospos_vs_negpos_genes, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
summary(GO_up_pospos_vs_negpos_results)
plot(barplot(GO_up_pospos_vs_negpos_results, showCategory = 10))

#pospos_vs_negpos decreased genes
GO_down_pospos_vs_negpos_genes = rownames(sig_pospos_vs_negpos[sig_pospos_vs_negpos$log2FoldChange < 1,])
GO_down_pospos_vs_negpos_results = enrichGO(gene = GO_down_pospos_vs_negpos_genes, OrgDb = "org.Hs.eg.db", keyType = "ENSEMBL", ont = "BP")
summary(GO_down_pospos_vs_negpos_results)
#plot(barplot(GO_down_pospos_vs_negpos_results, showCategory = 10))
##NO SIG PATHWAYS


```

# Autoencoder dimension reduction (IN PROGRESS)

```{r}
#try an autoencoder#


```

# WGCNA

```{r}

powers = seq(from = 1, to = 50, by = 1)

sft = pickSoftThreshold(df_transposed, 
                  powerVector = powers,
                  networkType = "signed",
                  )

ggplot(sft$fitIndices, aes(x=Power, y=SFT.R.sq)) +
  geom_point()

ggplot(sft$fitIndices, aes(x=Power, y=mean.k)) +
  geom_point()

```

